drivers/hmis/app/models/hmis/form/definition.rb

```diff
@@ -86,7 +88,21 @@ class Hmis::Form::Definition < ::GrdaWarehouseBase
     where(identifier: instance_scope.pluck(:definition_identifier))
   end
 
-  scope :with_role, ->(role) { where(role: role) }
+  scope :for_service_type, ->(service_type) do
+    instance_scope = Hmis::Form::Instance.none
+
+    base_scope = Hmis::Form::Instance.joins(:definition)
+    [
+      base_scope.for_service_type(service_type.id),
+      base_scope.for_service_category(service_type.custom_service_category_id),
+    ].each do |scope|
+      next if instance_scope.present?
+
+      instance_scope = scope unless scope.empty?
+    end
```

Comment:
Not sure what's going on here, but doesn't calling `scope.empty?` fire an active record call each time?  This seems like it might be fine given the number of items in the array, but generally, not great in a scope....
Is the goal to determine which of the two scopes has records?

Maybe:
```suggestion
    base_scope = Hmis::Form::Instance.joins(:definition)
    instance_scope = [
      base_scope.for_service_type(service_type.id),
      base_scope.for_service_category(service_type.custom_service_category_id),
    ].detect do |scope|
      scope.exists?
    end
```

Reply:
Yeah, I copied the logic from the `for_project` scope above. I'll see if I can simplify them

Reply:
Thanks for your suggestion, it worked exactly. Much easier to understand https://github.com/greenriver/hmis-warehouse/pull/3068/commits/f6ad25e5634311abb6f9c23b03f37c80ea93ca19